### PR TITLE
Enable no_std support.

### PR DESCRIPTION
**Pull request by Jezza**
`domain`, while offering a std feature, doesn't compile without std, regardless of the feature.

I've marked this PR as a WIP, as there's still some open questions.

We do seemingly need rng in a couple of places.
There is some support for non-os based rng in `rand`, but we'll probably need to use some static variable to hold the value.
There's also the question of seeding the thing.

### CODE REVIEW

File: src/base/header.rs [hunk 1]
```diff
@@ -407,15 +407,15 @@ impl FromStr for Flags {
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         let mut flags = Flags::new();
-        for token in s.to_uppercase().split(' ') {
+        for token in s.split(' ') {
```
Comment (Jezza) [inline src/base/header.rs:410]: Not exactly the prettiest thing, but to_uppercase() does allocate, and we only need to compare some basic ASCII characters.
We could probably be smarter about it, and just uppercase the ascii range, which is doable, but this seems to just be simple enough to not have to worry about it.

**Comment by Jezza** [comment]
There's two remaining spots that are currently blocking me.
I do see the try operator in use, and I'm pretty sure that's also not going to work.
So, we might have to build a `try` macro.

`src/base/header.rs:130:29`
image

`src/base/opt/rfc7830.rs:85:13`
image

**Comment by Jezza** [comment]
Almost done.

There's one last challenge.
How to pick a seed for RNG...

**Comment by partim** [comment]
Thanks for the PR!

On the random generator issue: Given that `Header::set_random_id` is only used by the resolver which requires the std feature, anyway, and a convenience method in `MessageBuilder`, I think it is okay to simply only include the method if std is set and make rand an optional dependency.

**Comment by Jezza** [comment]
Awesome, that just leaves `rfc7830.rs`, which uses it to generate random bytes for padding.

**Comment by partim** [comment]
The RFC actually says that zeros SHOULD be used. So I think making `PaddingMode` and `Padding::mode` std-only, rename the current `Padding::new` and `Padding::push` to `{new,push}_with_mode`, and add new `new` and `push` methods without mode is probably the best way forward.

**Comment by Jezza** [comment]
> Thanks for the PR!
> 
> On the random generator issue: Given that `Header::set_random_id` is only used by the resolver which requires the std feature, anyway, and a convenience method in `MessageBuilder`, I think it is okay to simply only include the method if std is set and make rand an optional dependency.

Should the whole function `request_axfr` be dependent on std, or should I just remove the "sets random id" part of the function.

**Comment by partim** [comment]
I think making the whole function depending on std is fine, at least for now. We can revisit it later. Perhaps stick on comment on it, stating that this requires std because of the random ID.

**Comment by Jezza** [comment]
Alrighty, changes are done.

I've gone with marking `PaddingMode::Random` as `feature = "std"`, as no point making whole types vanish.  
The `new` and `push` just use `PaddingMode::Zero` by default.  
I've added the `{new,push}_with_mode`, so that should cover any particular use cases remaining.

File: Cargo.toml [hunk 2]
```diff
@@ -39,7 +38,7 @@ master      = ["std", "bytes", "chrono"]
 resolv      = ["bytes", "futures", "smallvec", "std", "tokio", "libc"]
 resolv-sync = ["resolv", "tokio/rt"]
 sign        = ["std"]
-std         = []
+std         = ["rand", "rand/std", "rand/std_rng"]
```
Comment (partim) [inline Cargo.toml:41]: Do we still want the additional _rand_ features?

**Comment by partim** [comment]
I fixed the Clippy errors in main. Could you merge from main to hopefully make the CI go green?

**Comment by bugadani** [comment]
Without knowing any of the details, would it be possible to separate the rand and std features? rand does not (yet?) support the espidf OS, which means domain can only be used for ESP32 targets without `std` support, which makes working with domain a bit less comfortable.

**Comment by Jezza** [comment]
Hm, yeah, actually, would be easy enough.

**Comment by Jezza** [comment]
Now you can do something like: `default-features = false, features = ["std"]`.

**Comment by partim** [comment]
Note that the resolver needs random. I suspect that’s what you’d want to use on the ESP32? So perhaps we do need a work-around after all.

**Comment by Jezza** [comment]
I don't think so.

I'm using it for an esp32 as well, and `domain` is only used in the `experimental` feature.
After I made these changes, it builds fine, so it's not using any `random` feature.

https://github.com/esp-rs/embedded-svc/blob/master/src/utils/captive.rs#L6-L13

**Comment by bugadani** [comment]
Well it works perfectly for me, too, so I appreciate the change :)

**Comment by Jezza** [comment]
> Note that the resolver needs random. I suspect that’s what you’d want to use on the ESP32? So perhaps we do need a work-around after all.

It does look like `embedded-svc` does offer a DNS server impl, but that's behind its own `std` feature flag.

**Comment by partim** [comment]
Ok. In this case, can you add `random` to the features required by `resolv`? That would be the last thing before the PR can be merged.

**Comment by Jezza** [comment]
Oh, yeah, of course.

**Comment by partim** [comment]
Once again, thank you for the PR! There is likely going to be release 0.7.0 later this week.

**Pull request merged**